HDR images don't get constrained when Safari goes to background

mwyrzykowski opened the pull request:
#### 10635ec82dfc5ca52f9f2ad3a6876ebcfb8e97f8
<pre>
HDR images don&apos;t get constrained when Safari goes to background
<a href="https://bugs.webkit.org/show_bug.cgi?id=294092">https://bugs.webkit.org/show_bug.cgi?id=294092</a>
<a href="https://rdar.apple.com/151892044">rdar://151892044</a>

Reviewed by Matt Woodrow.

Constrain the dynamic range on backgrounding and repaint so
all images are rendered with the correct headroom.

* Source/WebCore/Modules/WebGPU/GPUCompositorIntegration.cpp:
(WebCore::GPUCompositorIntegration::updateContentsHeadroom):
Forward to backing.

* Source/WebCore/Modules/WebGPU/GPUCompositorIntegration.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUCompositorIntegrationImpl.cpp:
(WebCore::WebGPU::CompositorIntegrationImpl::updateContentsHeadroom):
Tell GPUP to update IOSurface headroom.

* Source/WebCore/Modules/WebGPU/Implementation/WebGPUCompositorIntegrationImpl.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUCompositorIntegration.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::screenPropertiesDidChange):
(WebCore::Document::addScreenPropertiesChangedObserver):
Add a notification for client&apos;s to subscribe and respond to screen property changes.

* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::createContextWebGPU):
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::getContext):
Pass Document to the GPUCanvasContext.

* Source/WebCore/html/canvas/GPUCanvasContext.cpp:
(WebCore::GPUCanvasContext::create):
* Source/WebCore/html/canvas/GPUCanvasContext.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
(WebCore::GPUCanvasContext::create):
(WebCore::GPUCanvasContextCocoa::create):
(WebCore::GPUCanvasContextCocoa::GPUCanvasContextCocoa):
(WebCore::GPUCanvasContextCocoa::updateContentsHeadroom):
(WebCore::GPUCanvasContextCocoa::configure):
Subscribe to screen property changes and forward to the backing as needed.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::screenPropertiesDidChange):
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::setContentHeadroom const):
Add member function for updating IOSurface headroom.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.cpp:
(WebKit::RemoteCompositorIntegration::updateContentsHeadroom):
Forward to implementation.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.messages.in:
Add message to update IOSurface headroom.

* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerNotificationObservers):
(WebKit::WebProcessPool::screenPropertiesChanged):
Backgrounding results in screen properties being changed, effectively.

* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::applicationShouldSuppressHDR):
(WebKit::setDynamicRangeLimitRecursive): Deleted.
(WebKit::setDynamicRangeLimit): Deleted.
No longer needed.

* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCompositorIntegrationProxy.cpp:
(WebKit::WebGPU::RemoteCompositorIntegrationProxy::updateContentsHeadroom):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCompositorIntegrationProxy.h:
Forward message to GPUP.

* LayoutTests/fast/webgpu/regression/edr-triangle-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/edr-triangle.html: Added.
Add EDR triangle test.

Canonical link: <a href="https://commits.webkit.org/296453@main">https://commits.webkit.org/296453@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/778bb515ae079017fc9a8886dc4dcf69c1d91bc9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108588 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28249 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18673 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113796 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58979 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110551 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28938 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36803 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82479 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111536 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22971 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97809 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62916 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22387 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15948 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58510 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92340 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15999 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116917 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35642 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26276 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/91503 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36015 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94080 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91308 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23260 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36204 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13963 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35543 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41077 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35254 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38600 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36932 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
```diff
@@ -677,6 +681,12 @@ static IntSize computeMaximumSurfaceSize()
+void IOSurface::setContentHeadroom(float headroom) const
+{
+    ASSERT(m_surface.get());
+    IOSurfaceSetValue(m_surface.get(), kIOSurfaceContentHeadroom, (CFTypeRef)@(headroom));
```
@mattwoodrow this tone maps the canvas context

mattwoodrow commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1240,6 +1248,15 @@ void setLockdownModeEnabledGloballyForTesting(std::optional<bool> enabledForTest
+    if (m_suppressHDR) {
+        for (auto& properties : screenProperties.screenDataMap.values()) {
+#if HAVE(SUPPORT_HDR_DISPLAY)
+            auto maxHeadroom = std::min(std::max(properties.maxEDRHeadroom * .5f, 1.5f), 2.f);
```
Can we use `std::clamp`, and some named constants for the upper/lower bounds. I think it'd be worth a comment about where the values came from too.

mattwoodrow commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1240,6 +1248,15 @@ void setLockdownModeEnabledGloballyForTesting(std::optional<bool> enabledForTest
+#if HAVE(SUPPORT_HDR_DISPLAY)
+            auto maxHeadroom = std::min(std::max(properties.maxEDRHeadroom * .5f, 1.5f), 2.f);
+            properties.currentEDRHeadroom = std::min(properties.currentEDRHeadroom, maxHeadroom);
+            properties.maxEDRHeadroom = std::min(properties.maxEDRHeadroom, maxHeadroom);
```
I think this one can just write `maxHeadroom`, no min required (or do it directly in the initial line, without the temp var).

mattwoodrow commented on Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
```diff
@@ -681,6 +681,12 @@ static IntSize computeMaximumSurfaceSize()
     IOSurfaceSetValue(m_surface.get(), kIOSurfaceColorSpace, colorSpaceProperties.get());
 }
 
+void IOSurface::setContentHeadroom(float headroom) const
```
This is a duplicate of `setContentEDRHeadroom` below.

mattwoodrow commented on Source/WebCore/html/OffscreenCanvas.cpp:
```diff
@@ -232,11 +232,11 @@ ExceptionOr<std::optional<OffscreenRenderingContext>> OffscreenCanvas::getContex
             Ref scriptExecutionContext = *this->scriptExecutionContext();
             if (RefPtr globalScope = dynamicDowncast<WorkerGlobalScope>(scriptExecutionContext)) {
                 if (auto* gpu = globalScope->protectedNavigator()->gpu())
-                    m_context = GPUCanvasContext::create(*this, *gpu);
```
Poor workers :(

Worth filing a bug with a fixme, to get worker contexts notified of the change?

mwyrzykowski replied:
I am wondering about this, isn't the worker displayed in a Document via an OffscreenCanvas? But I was not sure how to obtain the Document the OffscreenCanvas is displayed in, that was not immediately obvious to me

mattwoodrow replied:
It’s usually in a separate thread, so we cant access the Document. We’d need a threadsafe way to notify these observers

mwyrzykowski replied:
Let me see if this is easily resolvable as it would be good to have this behavior for workers too

mwyrzykowski replied:
Mitigated this https://github.com/WebKit/WebKit/pull/46868/files#r2155316667 by taking the max headroom from any connected display prior to getting the OffscreenCanvas contents

mwyrzykowski commented on Source/WebCore/platform/graphics/cocoa/IOSurface.h:
```diff
@@ -218,6 +218,7 @@ class IOSurface final {
     WEBCORE_EXPORT static void setOwnershipIdentity(IOSurfaceRef, const ProcessIdentity&);
 
     RetainPtr<CGContextRef> createCompatibleBitmap(unsigned width, unsigned height);
+    void setContentHeadroom(float) const;
```
will remove this

mwyrzykowski commented on Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
```diff
@@ -204,6 +232,16 @@ static GPUIntegerCoordinate getCanvasHeight(const GPUCanvasContext::CanvasType&
     if (!m_configuration)
         return canvasBase().buffer();
 
+    // FIXME: https://bugs.webkit.org/show_bug.cgi?id=294654 - OffscreenCanvas may not reflect the display the OffscreenCanvas is displayed on during background / resume
```
Offscreen canvas uses the max headroom from any connected display

mwyrzykowski commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -536,8 +536,11 @@ static bool hasValue(const ImageBufferBackendHandle& backendHandle)
-                auto surface = WebCore::IOSurface::createFromSendRight(WTFMove(machSendRight));
-                contents = surface ? surface->asCAIOSurfaceLayerContents() : nil;
+                if (auto surface = WebCore::IOSurface::createFromSendRight(WTFMove(machSendRight))) {
+                    if (surface->hasFormat(WebCore::IOSurface::Format::RGBA16F) && !surface->contentEDRHeadroom())
```
this change was needed @mattwoodrow

mwyrzykowski committed:
HDR images don't get constrained when Safari goes to background
https://bugs.webkit.org/show_bug.cgi?id=294092
rdar://151892044

Reviewed by Matt Woodrow.

Constrain the dynamic range on backgrounding and repaint so
all images are rendered with the correct headroom.

* Source/WebCore/Modules/WebGPU/GPUCompositorIntegration.cpp:
(WebCore::GPUCompositorIntegration::updateContentsHeadroom):
Forward to backing.

* Source/WebCore/Modules/WebGPU/GPUCompositorIntegration.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUCompositorIntegrationImpl.cpp:
(WebCore::WebGPU::CompositorIntegrationImpl::updateContentsHeadroom):
Tell GPUP to update IOSurface headroom.

* Source/WebCore/Modules/WebGPU/Implementation/WebGPUCompositorIntegrationImpl.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUCompositorIntegration.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::screenPropertiesDidChange):
(WebCore::Document::addScreenPropertiesChangedObserver):
Add a notification for client's to subscribe and respond to screen property changes.

* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::createContextWebGPU):
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::getContext):
Pass Document to the GPUCanvasContext.

* Source/WebCore/html/canvas/GPUCanvasContext.cpp:
(WebCore::GPUCanvasContext::create):
* Source/WebCore/html/canvas/GPUCanvasContext.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
(WebCore::GPUCanvasContext::create):
(WebCore::GPUCanvasContextCocoa::create):
(WebCore::GPUCanvasContextCocoa::GPUCanvasContextCocoa):
(WebCore::GPUCanvasContextCocoa::updateContentsHeadroom):
(WebCore::GPUCanvasContextCocoa::configure):
Subscribe to screen property changes and forward to the backing as needed.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::screenPropertiesDidChange):
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::setContentHeadroom const):
Add member function for updating IOSurface headroom.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.cpp:
(WebKit::RemoteCompositorIntegration::updateContentsHeadroom):
Forward to implementation.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.messages.in:
Add message to update IOSurface headroom.

* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerNotificationObservers):
(WebKit::WebProcessPool::screenPropertiesChanged):
Backgrounding results in screen properties being changed, effectively.

* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::applicationShouldSuppressHDR):
(WebKit::setDynamicRangeLimitRecursive): Deleted.
(WebKit::setDynamicRangeLimit): Deleted.
No longer needed.

* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCompositorIntegrationProxy.cpp:
(WebKit::WebGPU::RemoteCompositorIntegrationProxy::updateContentsHeadroom):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCompositorIntegrationProxy.h:
Forward message to GPUP.

* LayoutTests/fast/webgpu/regression/edr-triangle-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/edr-triangle.html: Added.
Add EDR triangle test.

Canonical link: https://commits.webkit.org/296453@main
